### PR TITLE
Add and remove invitation to and from group

### PR DESCRIPTION
**Pull request by duanxianbo**
image

image

updated:

image

### CODE REVIEW

File: src/client/graphql/fragment/invitation.graphql [hunk 1]
```diff
@@ -0,0 +1,24 @@
+
```
Comment (naturetime10) [inline src/client/graphql/fragment/invitation.graphql]:
```suggestion
```

File: src/client/graphql/fragment/invitation.graphql [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+fragment InvitationFragment on Invitation {
+    id
+    createdAt
+    updatedAt
+    expireAt
+    code
+    status
+    receiverFirstName
+	receiverLastName
+	receiverEmail
```
Comment (naturetime10) [inline src/client/graphql/fragment/invitation.graphql]:
```suggestion
    receiverLastName
    receiverEmail
```

File: src/client/invitation.client.ts [hunk 3]
```diff
@@ -5,6 +5,17 @@ import {
...
+import getInvitationsByTeamId from './graphql/query/getInvitationsByTeamId.graphql';
+import createInvitation from './graphql/mutation/createInvitation.graphql';
+import updateInvitation from './graphql/mutation/updateInvitation.graphql';
+import acceptInvitation from './graphql/mutation/acceptInvitation.graphql';
+import deleteInvitation from './graphql/mutation/deleteInvitation.graphql';
+import declineInvitation from './graphql/mutation/declineInvitation.graphql';
+import addInvitationToTeamMemberGroup from './graphql/mutation/addInvitationToTeamMemberGroup.graphql';  
+import removeInvitationFromTeamMemberGroup from './graphql/mutation/removeInvitationFromTeamMemberGroup.graphql';
+
+import { print } from 'graphql';
```
Comment (naturetime10) [inline src/client/invitation.client.ts]: I think this should be put above all other imports.

File: src/client/invitation.client.ts [hunk 4]
```diff
@@ -214,25 +112,52 @@ export class InvitationClient {
...
+            })
+            .then(
+                (payload: any) =>
+                    payload?.removeInvitationFromTeamMemberGroup,
+            );
+    }
 }
+
+
+
```
Comment (naturetime10) [inline src/client/invitation.client.ts]:
```suggestion
```

File: src/components/internal/Modals/TeamSettingsModal.component.module.scss [hunk 5]
```diff
@@ -437,3 +437,51 @@ $tableRowGap: 12px;
         }
     }
 }
+
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.module.scss:450]:
```suggestion
```

File: src/components/internal/Modals/TeamSettingsModal.component.module.scss [hunk 6]
```diff
@@ -437,3 +437,51 @@ $tableRowGap: 12px;
...
+        .Name {
+            align-items: center;
+            display: flex;
+            flex: 1;
+            font-weight: bold;
+            margin-left: 10px;
+            white-space: nowrap;
+        }
+}
+
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.module.scss]:
```suggestion
```

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 7]
```diff
@@ -35,11 +35,14 @@ import { UserProfileUI } from '../UserProfile';
 import { CreateTeamGroupModalComponent } from 'components/pages/Marketplace/Modals/CreateTeamGroupModal.component';
 import { CreateTeamMemberGroupModalComponent } from './CreateTeamMemberGroupModal.component';
 import { User } from '@core/entity/user';
-import { SelectUserPopupComponent } from '../SelectUserPopup.component';
+import { SelectEntityPopupComponent } from '../SelectEntityPopupComponent.component';
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
import { SelecItemPopupComponent } from '../SelectItemPopupComponent.component';
```

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 8]
```diff
@@ -240,32 +245,40 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
+            (invitation) => !groupInvitationIds.has(invitation.id),
+        );
+
+        const addUserOrInvitationDisabled =
+            !availableMemberUsers?.length && !availableGroupInvitations.length;
+        const isEditing = this.state.editingTeamMemberGroupInput[group.id];
+
+        const selectedUsersAndInvitationsSize =
+            (this.state.selectedGroupMembers[group.id]?.size || 0) +
+            (this.state.selectedGroupInvitations[group.id]?.size || 0);
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]: Can we use a separate variables for those 2 sizes?

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 9]
```diff
@@ -274,185 +287,306 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
-                                                        showFollower,
-                                                },
-                                            )}`}
-                                            onClick={this.handleToggleUserPopup(
-                                                !availableMemberUsers?.length,
-                                                toggleContainer,
+                                        <MaterialIconUI>remove</MaterialIconUI>
+                                    </div>
+                                </TooltipUI>
+                                <SelectEntityPopupComponent
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
                                <SelectItemPopupComponent
```

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 10]
```diff
@@ -274,185 +287,306 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
+                                                                >
+                                                                    {getUserShortName(
+                                                                        invitation.receiverFirstName,
+                                                                        invitation.receiverLastName,
+                                                                    )}
+                                                                </div>
+                                                            </div>
+                                                        ),
+                                                    )}
+                                                </>
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]: Maybe use a separate function?

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 11]
```diff
@@ -464,7 +598,9 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
         const teamMemberGroups = this.state.team?.memberGroups || [];
         const allMemberUsers =
             this.state.team?.members?.map((member) => member.user) || [];
-
+        const allNonAcceptedInvitations = this.state.invitations.filter(
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
        const allNotAcceptedInvitations = this.state.invitations.filter(
```

Comment (naturetime10) [reply src/components/internal/Modals/TeamSettingsModal.component.tsx]: I think we should use pending invitation instead. If invitation is declined, hide from this list.

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 12]
```diff
@@ -599,6 +734,78 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
+                    }`}
+                >
+                    {invitation.status}
+                </div>
+                <div className={styles.MemberName}>
+                    <span>{invitation.receiverFirstName}</span>
+                    &nbsp;
+                    <span>{invitation.receiverLastName}</span>
+                </div>
+
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
```

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 13]
```diff
@@ -813,26 +1024,66 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
+                );
+            });
+
+            selectedGroupInvitations?.forEach(async (invitationId) => {
+                await this.stateSyncer.removeInvitationFromTeamMemberGroup(
+                    invitationId,
+                    groupId,
+                );
+            });
+
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
```

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 14]
```diff
@@ -813,26 +1024,66 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
-            },
-        });
-    };
+            selectedGroupMembers?.forEach(async (userId) => {
+                await this.stateSyncer.removeUserFromTeamMemberGroup(
+                    groupId,
+                    userId,
+                );
+            });
+
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
```

File: src/components/internal/SelectEntityPopupComponent.component.module.scss [hunk 15]
```diff
@@ -1,4 +1,4 @@
-.Users {
+.Entities {
```
Comment (naturetime10) [inline src/components/internal/SelectEntityPopupComponent.component.module.scss]:
```suggestion
.Items {
```

File: src/components/internal/SelectEntityPopupComponent.component.tsx [hunk 16]
```diff
@@ -1,32 +1,23 @@
 import { RelativeLayout, RelativeLayoutContainerUI } from '@lib/layout';
 import { FC } from 'react';
-import styles from './SelectUserPopup.component.module.scss';
+import styles from './SelectEntityPopupComponent.component.module.scss';
```
Comment (naturetime10) [inline src/components/internal/SelectEntityPopupComponent.component.tsx]:
```suggestion
import styles from './SelectItemPopupComponent.component.module.scss';
```

Comment (naturetime10) [reply src/components/internal/SelectEntityPopupComponent.component.tsx]: Can we move this after line 4?

File: src/components/internal/SelectEntityPopupComponent.component.tsx [hunk 17]
```diff
@@ -1,32 +1,23 @@
...
 import { MaterialIconUI } from '@lib/ui/MaterialIcon';
-import { SelectUserListUI } from './SelectUserList';
-import { User } from '@core/entity/user';
 
 interface Props {
     relativeLayout: RelativeLayout;
-    users: User[];
-    onUserSelect: (userId: number) => void;
+    objectList: (closeContainer: () => void) => JSX.Element;
```
Comment (naturetime10) [inline src/components/internal/SelectEntityPopupComponent.component.tsx]:
```suggestion
    items: (closeContainer: () => void) => ReactElement;
```

File: src/components/internal/SelectEntityPopupComponent.component.tsx [hunk 18]
```diff
@@ -1,32 +1,23 @@
...
-        };
+export const SelectEntityPopupComponent: FC<Props> = (props) => {
     return (
         <RelativeLayoutContainerUI
             relativeLayout={props.relativeLayout}
             zOffset={props.zOffset}
             renderFollower={({ closeContainer }) => {
                 return (
-                    <div className={styles.Users}>
+                    <div className={styles.Entities}>
```
Comment (naturetime10) [inline src/components/internal/SelectEntityPopupComponent.component.tsx]:
```suggestion
                    <div className={styles.Items}>
```

File: src/components/internal/SelectEntityPopupComponent.component.tsx [hunk 19]
```diff
@@ -35,13 +26,8 @@ export const SelectUserPopupComponent: FC<Props> = (props) => {
...
                             </div>
                         </div>
-                        <div className={styles.UserList}>
-                            <SelectUserListUI
-                                users={props.users}
-                                onSelectUser={handleUserSelectClick(
-                                    closeContainer,
-                                )}
-                            />
+                        <div className={styles.EntityList}>
```
Comment (naturetime10) [inline src/components/internal/SelectEntityPopupComponent.component.tsx]:
```suggestion
                        <div className={styles.Items}>
```

File: src/components/pages/Sprints.component.tsx [hunk 20]
```diff
@@ -45,7 +45,8 @@ import { KanbanUI } from '../internal/Kanban';
 import { UserProfileUI } from '../internal/UserProfile';
 import { getSprintName, getUserShortName } from '../internal/format';
 import styles from './Sprints.component.module.scss';
```
Comment (naturetime10) [inline src/components/pages/Sprints.component.tsx]: Can we move this after line 49?

File: src/components/pages/Sprints.component.tsx [hunk 21]
```diff
@@ -464,15 +466,21 @@ export class SprintsComponent extends Component<Props, State> {
                                                 </div>
                                             </TooltipUI>
                                         </div>
-                                        <SelectUserPopupComponent
+                                        <SelectEntityPopupComponent
```
Comment (naturetime10) [inline src/components/pages/Sprints.component.tsx]:
```suggestion
                                        <SelectItemPopupComponent
```

File: src/storage/graph/teamMemberGroup.node.ts [hunk 22]
```diff
@@ -46,4 +47,20 @@ export class TeamMemberGroupNode implements TeamMemberGroup {
...
+                (relation) => relation.groupId === this.teamMemberGroupState.id
+            )
+            .map(
+                (relation) =>
+                    new InvitationNode(
+                        this.localStore,
+                        appState.invitations[relation.invitationId],
+                    ),
+            )
+            .filter((invitation) => invitation.status !== 'ACCEPTED');
```
Comment (naturetime10) [inline src/storage/graph/teamMemberGroup.node.ts]:
```suggestion
            .filter((invitation) => invitation.status === 'PENDING');
```

File: src/storage/syncer/stateMutators.ts [hunk 23]
```diff
@@ -504,6 +505,22 @@ export function applyTaskAwaitingForRelationMutation(
...
+        case 'Create': {
+            const relation = TeamMemberGroupInvitationRelation.fromMutationPayload(mutation.payload);
+            currState.teamMemberGroupInvitationRelations = currState.teamMemberGroupInvitationRelations.concat(relation);
+            break;
+        }
+    }
+
+    return currState;
+}
+
```
Comment (naturetime10) [inline src/storage/syncer/stateMutators.ts]:
```suggestion
```

File: src/storage/syncer/stateMutators.ts [hunk 24]
```diff
@@ -1623,11 +1640,51 @@ export function mergeTeamMemberGroup(
...
 }
 
+export function mergeTeamMemberGroupInvitationRelation(
+    currAppState: AppState,
+    remoteTeamMemberGroupId: string,
+    remoteInvitationId: string,
+): AppState {
+    const teamMemberGroupId = toInt(remoteTeamMemberGroupId)!;
+    const invitationId = toInt(remoteInvitationId)!;
+
```
Comment (naturetime10) [inline src/storage/syncer/stateMutators.ts]:
```suggestion
```

File: src/storage/syncer/stateMutators.ts [hunk 25]
```diff
@@ -1623,11 +1640,51 @@ export function mergeTeamMemberGroup(
...
+    const hasTeamMemberGroupInvitationRelation =
+        currAppState.teamMemberGroupInvitationRelations.some(
+            (teamMemberGroupInvitationRelation) => {
+                return (
+                    teamMemberGroupInvitationRelation.groupId ===
+                        teamMemberGroupId &&
```
Comment (naturetime10) [inline src/storage/syncer/stateMutators.ts]:
```suggestion
                    teamMemberGroupInvitationRelation.groupId === teamMemberGroupId &&
```

File: src/storage/syncer/stateMutators.ts [hunk 26]
```diff
@@ -1623,11 +1640,51 @@ export function mergeTeamMemberGroup(
...
+                    teamMemberGroupInvitationRelation.invitationId ===
+                        invitationId
```
Comment (naturetime10) [inline src/storage/syncer/stateMutators.ts]:
```suggestion
                    teamMemberGroupInvitationRelation.invitationId === invitationId
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 27]
```diff
@@ -274,185 +287,295 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
+                                            <SelectUserListUI
+                                                users={
+                                                    availableMemberUsers || []
+                                                }
+                                                onSelectUser={this.handleAddUserToTeamMemberGroup(
+                                                    closeContainer,
+                                                    group.id,
+                                                )}
+                                            />
+
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by naturetime10** [comment]
Forgot to push some commits? Can we add a new screenshot for the user list popup?

**Comment by duanxianbo** [comment]
> Forgot to push some commits? Can we add a new screenshot for the user list popup?

done

File: src/components/internal/Modals/TeamSettingsModal.component.module.scss [hunk 28]
```diff
@@ -447,3 +447,36 @@ $tableRowGap: 12px;
         }
     }
 }
+.InvitationProfile {
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.module.scss:451]:
```suggestion

.InvitationProfile {
```

File: src/client/invitation.client.ts [hunk 29]
```diff
@@ -1,10 +1,19 @@
 import { GraphQLClient } from '@lib/network/GraphQL.client';
-
+import { print } from 'graphql';
```
Comment (naturetime10) [inline src/client/invitation.client.ts:1]: Can we move this above line 1?

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 30]
```diff
@@ -503,7 +636,9 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
         const teamMemberGroups = this.state.team?.memberGroups || [];
         const allMemberUsers =
             this.state.team?.members?.map((member) => member.user) || [];
-
+        const allNotAcceptedInvitations = this.state.invitations.filter(
+            (invitation) => invitation.status !== 'ACCEPTED',
+        );
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
        const allPendingInvitations = this.state.invitations.filter(
            (invitation) => invitation.status === 'PENDING',
        );
```

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 31]
```diff
@@ -515,18 +650,16 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
                     </div>
                 </div>
                 <div className={styles.Scrollable}>
-                    {this.renderTeamMemberGroup({
-                        groupName: 'All Members',
-                        groupMemberUsers: allMemberUsers,
-                    })}
+                    {this.renderAllTeamMemberGroup(
+                        allMemberUsers,
+                        allNotAcceptedInvitations,
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
                        allPendingInvitations,
```

File: src/components/internal/Modals/TeamSettingsModal.component.tsx [hunk 32]
```diff
@@ -515,18 +650,16 @@ export class TeamSettingsModalComponent extends Component<Props, State> {
...
-                            groupName: group.name,
-                            groupMemberUsers: group.members,
-                            teamMemberUsers: allMemberUsers,
-                            groupId: group.id,
-                            groupIndex: group.orderIndex,
-                        });
+                        return this.renderTeamMemberGroup(
+                            group,
+                            allMemberUsers,
+                            allNotAcceptedInvitations,
```
Comment (naturetime10) [inline src/components/internal/Modals/TeamSettingsModal.component.tsx]:
```suggestion
                            allPendingInvitations,
```

**Review by naturetime10 — APPROVED** [review]

File: src/client/invitation.client.ts [hunk 33]
```diff
@@ -214,25 +110,49 @@ export class InvitationClient {
...
+                variables: {
+                    invitationId,
+                    teamMemberGroupId,
+                },
+            })
+            .then(
+                (payload: any) =>
+                    payload?.addInvitationToTeamMemberGroup,
+            );
+
```
Comment (naturetime10) [inline src/client/invitation.client.ts]:
```suggestion
```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**